src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -2849,26 +2842,55 @@ public void watchFiles(File outputDirectory, File testOutputDirectory, final Thr
+                    // set to true regardless of success
+                    // this may need to be updated, but currently do not want to come into this section again
+                    registerClassDirAttempted = true;
+                    if (isMultiModuleProject()) {
+                        for (ProjectModule p : upstreamProjects) {
+                            if (shouldIncludeSources(p.getPackagingType())) {
+                                // register source classes directory for other modules
+                                if (p.getOutputDirectory().exists()) {
+                                    registerAll(p.getOutputDirectory().getCanonicalFile().toPath(), executor);
+                                    p.sourceOutputDirRegistered = true;
+                                }
                             }
                         }
                     }
+                    if (shouldIncludeSources(packagingType)) {
+                         // register source classes directory for main module
+                        if (this.outputDirectory.exists()) {
+                            registerAll(srcOutputPath, executor);
+                            sourceOutputDirRegistered = true;
+                        }
+                    }
+                }
+
+                // Generate features from source file changes
+                if (generateFeatures && !classesFailingToCompile()) { //do not run generate features if there are classes failing to compile
+                    boolean tryFeatureGeneration = false;
+                    if (recompileDependencies && lastChangeCompiled && !javaSourceClasses.isEmpty()) { // no class file tracking
+                        debug("Detected a change in the following class directories: " + javaSourceClasses);
+                        lastChangeCompiled = false;
+                        modifiedSrcBuildFile = null;
+                        tryFeatureGeneration = true;
+                        incrementGenerateFeatures();
+                    } else if (!recompileDependencies && !javaSourceClasses.isEmpty()) { // class file tracking
+                        debug("Detected a change in the following classes: " + javaSourceClasses);
+                        tryFeatureGeneration = true;
+                        incrementGenerateFeatures();
+                    }
+                    // TODO: look into whether or not we need this code block
+                    if (tryFeatureGeneration && !generatedFeaturesFile.exists()) {
+                        // run tests if generated-features.xml does not exist as there are no new features to install
+                        if (isMultiModuleProject()) {
+                            runTestThread(false, executor, -1, false, getAllBuildFiles());
+                        } else {
+                            runTestThread(false, executor, -1, false, false, buildFile);
+                        }
+                    }
```

Comment:
I think we still need this block of code. If dev mode starts and the server.xml already contains all of the features needed, we will not create a generated-features.xml file. This catches the case where `hotTests=true`, we want to run tests after feature generation finishes.